### PR TITLE
Adding type hints for mypy strict compatibility

### PR DESCRIPTION
**Pull request by unna97**
Adding type hints for mypy strict compatibility.

### CODE REVIEW

**Comment by unna97** [comment]
This pull request tries to solve issues as highlighted in issue: #501

**Comment by unna97** [comment]
@jwmueller  is the NumPy type as done in the file okay?

**Comment by jwmueller** [comment]
Thanks for your contribution @unna97!
We are in the midst of figuring this numpy typing out ourselves since we just dropped Python 3.6 support which was holding us back from proper numpy type annotations.

Our goal is to provide one reference file with proper numpy type annotations which contributors can reference to easily update the type annotations in any other file. The PR for this is here:  https://github.com/cleanlab/cleanlab/pull/587
So check that out and follow along our discussion; as soon as it's merged, you can update your PR to follow the same pattern.

**Comment by elisno** [comment]
> @jwmueller is the NumPy type as done in the file okay?

Hey @unna97, thanks for the PR!

It looks like you're on the right track with the numpy arrays 👍.

 You should modify the type parametrization like is done in the PR [that Jonas mentions](https://github.com/cleanlab/cleanlab/pull/585#issuecomment-1374204809).

The rule of thumb here is that changes in the PR should have fewer `mypy --strict` errors on a per-file basis, without the total number of errors going up for the entire package.

**Comment by elisno** [comment]
@unna97: Once you mark this PR as "Ready for review", feel free to send me a request for a review.

**Comment by unna97** [comment]
@elisno  @jwmueller  thank you so much for the answers & helpful docs/examples. Would you prefer that I make one PR for each file or a commit per file for solving mypy errors?

**Comment by elisno** [comment]
Good question!

Of those two options, a commit-per-file approach would be better.
You don't have to constrain yourself this way.
It's perfectly reasonable to use several commits per file.

**Comment by unna97** [comment]
Files with type issues:

- [ ]  [cleanlab/internal/token_classification_utils.py]
- [ ]  [cleanlab/multilabel_classification.py]
- [ ]  [setup.py]
- [ ]  [cleanlab\token_classification\summary.py]
- [ ]  [cleanlab\internal\validation.py]
- [ ]  [cleanlab\internal\util.py]

**Comment by elisno** [comment]
> Files with type issues:
> 
> * [ ]   [setup.py]
> * [ ]  [cleanlab/dataset.py]
> * [ ]  [cleanlab/rank.py]
> ...

Let's focus on the modules within `cleanlab/`.
You can skip `setup.py` and everything in `tests/` for now.

--- 

To make it easier to review and accept your contributions, I recommend that the work would be split into several PRs.

- [ ] Top-level cleanlab modules
  - `cleanlab/datasets.py`, `cleanlab/rank.py`, ...
- [ ] Internal modules
  - `cleanlab/internal/token_classification_utils.py`, ...
- [ ] Token classification sub-module
  - `cleanlab/token_classification/rank.py`, ...

It will be tricky to fix all type-hints in some of the modules, so just a partial fix will be greatly appreciated!

File: cleanlab/internal/token_classification_utils.py [hunk 1]
```diff
@@ -188,7 +194,7 @@ def merge_probs(probs: np.ndarray, maps: List[int]) -> np.ndarray:
     Returns
     ---------
     probs_merged:
-        np.array of shape ``(N, K')``, where `K` is the number of new classes. Probabilities are merged and
+        A 2D np.array of shape ``(N, K')``, where `K` is the number of new classes. Probabilities are merged and
```
Comment (elisno) [inline cleanlab/internal/token_classification_utils.py:197]:
```suggestion
        A 2D np.array of shape ``(N, K')``, where `K'` is the number of new classes. Probabilities are merged and
```

**Comment by elisno** [comment]
@unna97, I think this is a great start!
Is it all right with you if I review the current changes and merge them?

**Comment by elisno** [comment]
Going back to your question in https://github.com/cleanlab/cleanlab/pull/585#issuecomment-1374521400:

My line of thinking at the time was that you should rarely update multiple files (with unrelated changes) in a single commit when working on a PR. I'd like to change my answer to the former option and say that it's preferable if you make one PR for one file.
After all, a bird in the hand is worth two in the bush!

**Comment by unna97** [comment]
So, I should not split them into 3 PRs as mentioned in https://github.com/cleanlab/cleanlab/pull/585#issuecomment-1374600878

**Comment by unna97** [comment]
https://github.com/cleanlab/cleanlab/pull/585#issuecomment-1376549901

**Comment by elisno** [comment]
It's absolutely fine to make separate PRs for individual files. It will be easier to review.

**Review by elisno — APPROVED** [review]
Great stuff!
Thank you @unna97!

**Pull request merged**